src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -0,0 +1,563 @@
+            {
+                await ShutdownAsync(QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0).ConfigureAwait(false);
+                _api.StreamCloseDelegate?.Invoke(_ptr);
+            }
+
+            _handle.Free();
+            _api = null;
+
+            _disposed = true;
+        }
+
+        public override void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        ~MsQuicStream()
+        {
+            Dispose(false);
+        }
+
+        // Synchronous shutdown current does a graceful shutdown, which must go async
+        // Close can be done synchronously, but there is not guarantee that all data will be sent to the client
+        // We probably need to reconsider how to handle dispose/shutdown cases.
+        private void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_ptr != IntPtr.Zero)
+            {
+                ShutdownAsync(QUIC_STREAM_SHUTDOWN_FLAG.GRACEFUL, 0).GetAwaiter().GetResult();
+                _api.StreamCloseDelegate?.Invoke(_ptr);
+            }
+
+            _handle.Free();
+            _api = null;
+
+            _disposed = true;
+        }
+
+        private void EnableReceive()
+        {
+            uint status = _api.StreamReceiveSetEnabledDelegate(_ptr, enabled: true);
+        }
+
+        private void DisableReceive()
+        {
+            uint status = _api.StreamReceiveSetEnabledDelegate(_ptr, enabled: false);
+        }
+
+        internal static uint NativeCallbackHandler(
+           IntPtr stream,
+           IntPtr context,
+           StreamEvent connectionEventStruct)
+        {
+            var handle = GCHandle.FromIntPtr(context);
+            var quicStream = (MsQuicStream)handle.Target;
+
+            return quicStream.HandleEvent(ref connectionEventStruct);
+        }
+
+        private uint HandleEvent(ref StreamEvent evt)
+        {
+            uint status = MsQuicConstants.Success;
+
+            try
+            {
+                switch (evt.Type)
+                {
+                    case QUIC_STREAM_EVENT.START_COMPLETE:
+                        status = HandleStartComplete();
+                        break;
+                    case QUIC_STREAM_EVENT.RECV:
+                        {
+                            HandleEventRecv(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.SEND_COMPLETE:
+                        {
+                            status = HandleEventSendComplete(ref evt);
+                        }
+                        break;
+                    case QUIC_STREAM_EVENT.PEER_SEND_CLOSE:
+                        {
+                            status = HandleEventPeerSendClose();
+                        }
+                        break;
+                    // TODO figure out difference between SEND_ABORT and RECEIVE_ABORT
```

Comment:
PEER_SEND_ABORT = peer is aborting their send direction.
PEER_RECV_ABORT = peer is aborting their receive direction.